firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,324 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+      // need to rewrite it
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testGetResolvesToServerCacheWhenListenerIsAvailable()
+      throws DatabaseException, InterruptedException {
+    // To ensure that we don't read the cached results, we need a separate app.
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    db.setPersistenceEnabled(false);
+    long val = 34;
+    DatabaseReference node = db.getReference().push();
+
+    try {
+      await(node.setValue(val));
+      Semaphore semaphore = new Semaphore(0);
+      ValueEventListener listener =
+          new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+              assertEquals(val, snapshot.getValue());
+              semaphore.release();
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {}
+          };
+      node.addValueEventListener(listener);
+      IntegrationTestHelpers.waitFor(semaphore);
+      db.goOffline();
+      DataSnapshot snapshot = await(node.get());
+      assertEquals(val, snapshot.getValue());
+      node.removeEventListener(listener);
+      db.goOnline();
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testGetResolvesToCacheWhenOnlineAndChildLevelListener()
+      throws DatabaseException, InterruptedException {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    Semaphore semaphore = new Semaphore(0);
+    long val = 34;
+    String parentNodeKey = Objects.requireNonNull(writeDb.getReference().push().getKey());
+    DatabaseReference parentWriteNode = writeDb.getReference().child(parentNodeKey);
+    String childNodeKey = Objects.requireNonNull(parentWriteNode.push().getKey());
+    DatabaseReference childReadNode =
+        readDb.getReference().child(parentNodeKey).child(childNodeKey);
+    DatabaseReference childWriteNode = parentWriteNode.child(childNodeKey);
+    ValueEventListener listener =
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            semaphore.release();
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {
+            // no-op
+          }
+        };
+
+    try {
+      await(childWriteNode.setValue(val));
+      new ReadFuture(
```

Comment:
We could use `untilEquals` here.

Also, it's possible that the completion condition never gets called, so we should generally prefer to make assertions on the return value of `ReadFuture.timedGet()`. Check out some of the examples in `DataTest.java`.